src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -0,0 +1,1123 @@
+        .expect("statistics initialized")
+        .clone();
+
+    let mut write_op =
+        AsyncOperatorBuilder::new(format!("{} write_batches", operator_name), scope.clone());
+
+    let (mut output, output_stream) = write_op.new_output();
+
+    let mut descriptions_input = write_op.new_input(&batch_descriptions.broadcast(), Pipeline);
+    let mut desired_input = write_op.new_input(desired_stream, Pipeline);
+
+    // This operator accepts the current and desired update streams for a `persist` shard.
+    // It attempts to write out updates, starting from the current's upper frontier, that
+    // will cause the changes of desired to be committed to persist, _but only those also past the
+    // upper_.
+
+    let shutdown_button = write_op.build(move |_capabilities| async move {
+        // In-progress batches of data, keyed by timestamp.
+        let mut stashed_batches = BTreeMap::new();
+
+        // Contains descriptions of batches for which we know that we can
+        // write data. We got these from the "centralized" operator that
+        // determines batch descriptions for all writers.
+        //
+        // `Antichain` does not implement `Ord`, so we cannot use a `BTreeMap`. We need to search
+        // through the map, so we cannot use the `mz_ore` wrapper either.
+        #[allow(clippy::disallowed_types)]
+        let mut in_flight_batches = std::collections::HashMap::<
+            (Antichain<Timestamp>, Antichain<Timestamp>),
+            Capability<Timestamp>,
+        >::new();
+
+        // TODO(aljoscha): We need to figure out what to do with error results from these calls.
+        let persist_client = persist_clients
+            .open(persist_location)
+            .await
+            .expect("could not open persist client");
+
+        let mut write = persist_client
+            .open_writer::<SourceData, (), Timestamp, Diff>(
+                shard_id,
+                &format!("compute::persist_sink::write_batches {}", collection_id),
+                Arc::new(target_relation_desc),
+                Arc::new(UnitSchema),
+            )
+            .await
+            .expect("could not open persist shard");
+
+        // The current input frontiers.
+        let mut batch_descriptions_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+        let mut desired_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        // The frontiers of the inputs we have processed, used to avoid redoing work
+        let mut processed_desired_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+        let mut processed_descriptions_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        // A "safe" choice for the lower of new batches we are creating.
+        let mut operator_batch_lower = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        loop {
+            tokio::select! {
+                Some(event) = descriptions_input.next_mut() => {
```

Comment:
Doing this will win you an indentation level:

```suggestion
                Some(event) = descriptions_input.next_mut() => match event {
```

Reply:
oh wait, `select!` doesnt support this :(

Reply:
It does, we do it in a bunch of places. You just need to put a trailing comma because it's an expression like so: `match event { .. },`

Reply:
UGH the trailing comma, that must have been it